File: Backend/scrapers/scrapeAmazon.js
// const axios = require("axios");
// const cheerio = require("cheerio");

// async function scrapeAmazon(query) {
//   const searchUrl = `https://www.amazon.in/s?k=${encodeURIComponent(query)}`;
//   const headers = {
//     "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
//   };

//   try {
//     console.log("🔍 Searching Amazon for:", query);
//     const { data: searchHtml } = await axios.get(searchUrl, { headers });
//     const $ = cheerio.load(searchHtml);

//     const firstLink = $("a.a-link-normal.s-no-outline").first().attr("href");

//     if (!firstLink) {
//       console.log("❌ No product link found on search page");
//       return null;
//     }

//     const productUrl = `https://www.amazon.in${firstLink}`;
//     console.log("✅ Navigating to Amazon product page:", productUrl);

//     const { data: detailHtml } = await axios.get(productUrl, { headers });
//     const $$ = cheerio.load(detailHtml);

//     const price =
//       $$("span.a-price-whole").first().text().trim() ||
//       $$("span.a-offscreen").first().text().trim();

//     const image = $$("img#landingImage").attr("src");

//     const reviews = [];
//     $$("div[data-hook='review']").each((_, el) => {
//       const star = $$(el).find("i[data-hook='review-star-rating'] span").text().trim().charAt(0);
//       const text = $$(el).find("span[data-hook='review-body']").text().trim();
//       if (star && text) {
//         reviews.push("★".repeat(Number(star)) + ` "${text}"`);
//       }
//     });

//     console.log("📦 Amazon Price:", price);
//     console.log("🖼️ Amazon Image:", image);
//     console.log("📝 Amazon Reviews Count:", reviews.length);

//     return {
//       price: price || "Not available",
//       image,
//       buyLink: productUrl,
//       reviews
//     };
//   } catch (err) {
//     console.error("❌ Amazon scraping failed:", err.message);
//     return null;
//   }
// }

// module.exports = scrapeAmazon;


const axios = require("axios");
const cheerio = require("cheerio");

async function getAmazonProductDetails(query) {
  try {
    const searchUrl = `https://www.amazon.in/s?k=${encodeURIComponent(query)}`;
    const headers = {
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
      "Accept-Language": "en-US,en;q=0.9"
    };

    // Step 1: Get product link from search page
    const searchRes = await axios.get(searchUrl, { headers });
    const $search = cheerio.load(searchRes.data);

    let productLink = null;
    $search("a.a-link-normal.s-no-outline").each((_, el) => {
      const href = $search(el).attr("href");
      if (href && href.includes("/dp/")) {
        productLink = "https://www.amazon.in" + href;
        return false; // break loop
      }
    });

    if (!productLink) throw new Error("No product link found on Amazon.");

    // Step 2: Scrape product details from product page
    const productRes = await axios.get(productLink, { headers });
    const $ = cheerio.load(productRes.data);

    const name = $("#productTitle").text().trim();
    const image = $("#imgTagWrapperId img").attr("src") || $("img[data-old-hires]").attr("src");
    const reviews = [];

    $("div[data-hook='review']").each((_, el) => {
      const ratingText = $(el).find("i[data-hook='review-star-rating'] span").text().trim();
      const rating = parseFloat(ratingText);
      const body = $(el).find("span[data-hook='review-body'] span").text().trim();
      if (!isNaN(rating) && body) {
        reviews.push("★".repeat(Math.round(rating)) + ` "${body}"`);
      }
    });

    return {
      name,
      image,
      buyLink: productLink,
      reviews
    };
  } catch (err) {
    console.error("Amazon error:", err.message);
    return { error: "Amazon scraping failed." };
  }
}

module.exports = getAmazonProductDetails;

File: Backend/scrapers/scrapeAmazon.py
import sys
import json
import requests
from bs4 import BeautifulSoup
from pymongo import MongoClient
from bson import ObjectId

HEADERS = {
    'User-Agent': (
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
        'AppleWebKit/537.36 (KHTML, like Gecko) '
        'Chrome/90.0.4430.212 Safari/537.36'
    ),
    'Accept-Language': 'en-US,en;q=0.5'
}

def get_html(url):
    response = requests.get(url, headers=HEADERS)
    response.raise_for_status()
    return response.text

def get_soup(url):
    html = get_html(url)
    return BeautifulSoup(html, 'html.parser')

def scrape_amazon_details(url):
    soup = get_soup(url)
    reviews, names, ratings = [], [], []

    price_tag = soup.find("span", class_="a-price-whole")
    amazon_price = None
    if price_tag:
        try:
            amazon_price = int(float(price_tag.get_text(strip=True).replace(",", "")))
        except:
            amazon_price = None

    name_tags = soup.find_all("span", class_="a-profile-name")
    for tag in name_tags:
        names.append(tag.get_text())

    review_divs = soup.find_all("div", class_="a-expander-content reviewText review-text-content a-expander-partial-collapse-content")
    for div in review_divs:
        lines = div.get_text().split("\n")
        for line in lines:
            if line.strip():
                reviews.append(line.strip())

    rating_tags = soup.find_all("i", class_="review-rating")
    for tag in rating_tags:
        text = tag.get_text(strip=True)
        try:
            ratings.append(float(text.split()[0]))
        except:
            continue

    avg_rating = None
    avg_tag = soup.find("span", class_="a-icon-alt")
    if avg_tag:
        try:
            avg_rating = float(avg_tag.get_text(strip=True).split()[0])
        except:
            pass

    image_url = None
    img_tag = soup.find("img", id="landingImage")
    if img_tag and img_tag.get("src"):
        image_url = img_tag["src"]

    n = min(len(names), len(reviews), len(ratings))
    scraped_reviews = []
    for i in range(n):
        scraped_reviews.append({
            "username": names[i],
            "review": reviews[i],
            "rating": ratings[i],
            "source": "amazon"
        })

    return scraped_reviews, avg_rating, amazon_price, image_url

# Run the scraper and update DB
if __name__ == "__main__":
    try:
        product_id = sys.argv[1]
        url = sys.argv[2]
    except IndexError:
        print(json.dumps({
            "success": False,
            "message": "Missing product ID or URL"
        }))
        sys.exit(1)

    reviews, avg_rating, price, image_url = scrape_amazon_details(url)

    try:
        client = MongoClient("mongodb+srv://vishal10992021:FJTBWV98N1Gk05dG@cluster0.nchvnmv.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0")
        db = client["test"]
        products = db["products"]

        update = {}
        if price is not None:
            update["prices.amazon"] = price
        if image_url:
            update["imageUrl"] = image_url
        if avg_rating:
            update["amazonAvgRating"] = avg_rating

        update_query = {"$set": update}

        if reviews:
            update_query["$push"] = {
                "reviews.amazon": {
                    "$each": [
                        {
                            "username": r["username"],
                            "comment": r["review"],
                            "rating": r["rating"],
                        }
                        for r in reviews
                    ]
                }
            }

        result = products.update_one(
            { "_id": ObjectId(product_id) },
            update_query
        )

        print(json.dumps({
            "success": True,
            "matched": result.matched_count,
            "modified": result.modified_count,
            "price": price,
            "image": image_url,
            "num_reviews": len(reviews)
        }))

    except Exception as e:
        print(json.dumps({
            "success": False,
            "message": f"MongoDB update failed: {str(e)}"
        }))
        sys.exit(1)
    finally:
        client.close()
